scheduler/src/cook/impersonation.clj

```diff
@@ -0,0 +1,87 @@
+;;
+;; Copyright (c) Two Sigma Open Source, LLC
+;;
+;; Licensed under the Apache License, Version 2.0 (the "License");
+;; you may not use this file except in compliance with the License.
+;; You may obtain a copy of the License at
+;;
+;;  http://www.apache.org/licenses/LICENSE-2.0
+;;
+;; Unless required by applicable law or agreed to in writing, software
+;; distributed under the License is distributed on an "AS IS" BASIS,
+;; WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+;; See the License for the specific language governing permissions and
+;; limitations under the License.
+;;
+(ns cook.impersonation
+  "Support for services impersonating a user, performing Cook actions the user's behalf."
+  (:require [clojure.string]
+            [clojure.tools.logging :as log]
+            [cook.util]
+            [ring.util.response :refer (header status response)])
+  (:import (clojure.lang Keyword)))
+
+(defn- unrestricted-access
+  "Denotes that an impersonator may perform any operation on a given object type."
+  [_]
+  true)
+
+(defn- getter-access
+  "Denotes that an impersonator may only perform gets on a given object type."
+  [verb]
+  (= :get verb))
+
+(def object-type->verb->impersonatable?
+  "Specification of operations that authorized impersonators may perform on behalf of another user."
+  {:job unrestricted-access
+   :quota getter-access
+   :share getter-access
+   :usage getter-access})
+
+(defn impersonation-authorized-wrapper
+  "This function is meant to wrap the authorization/is-authorized? function.
+
+  The resulting function determines whether an authorized impersonator
+  can perform the given operation on the given object, on behalf of the given user.
+
+  Returns true if allowed, else false."
+  [is-authorized-fn settings]
+  (fn is-impersonatable?
+    [^String user
+     ^Keyword verb
+     ^String impersonator
+     {:keys [owner item] :as object}]
+    (log/debug "[is-impersonatable?] Checking whether user" impersonator
+               "may impersonate user" user
+               "performing" verb "on" (str object) "...")
+    (log/debug "[is-impersonatable?] Settings are:" settings)
+    (and (or (nil? impersonator)
+             (if-let [verb->impersonatable? (object-type->verb->impersonatable? item)]
+               (verb->impersonatable? verb)
+               false))
+         (is-authorized-fn settings user verb impersonator object))))
+
+(defn create-impersonation-middleware
+  "Provides user-impersonation middleware.
+  Uses the impersonators set-argument to verify impersonation authorization."
+  [impersonators]
+  (fn impersonation-middleware [h]
+    (if-let [impersonators-set (some-> impersonators seq set)]
+      (fn impersonation-wrapper [{user :authorization/user :as req}]
```

Comment:
Are you saying that lifting `impersonation-wrapper` to a top-level def would get rid of the need for those comments? Or make them less smelly? Or are those two separate comments?

I was thinking about introducing an `if-not-let` macro so I could move the simple error cases up to the front, but I wasn't sure that was worth it. I'm also not convinced that lifting that function (and then having to explicitly pass the arguments that are currently closed-over) helps either.

Reply:
The gist of the comment was that it seems like the comments are necessary because the function is long and deeply nested enough it's not clear what the `if` was when you get to the `else`. I was hoping there was some reasonable way to break this up or lift out some of the content to make it more clear. If it's too tied together to split up, then I think the comments are helpful.

Reply:
OK, sounds like you don't like the macro idea. Leaving this as-is.